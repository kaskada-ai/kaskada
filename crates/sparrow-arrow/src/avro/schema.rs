use super::Error;
use arrow::datatypes::{DataType, Field, IntervalUnit, Schema, TimeUnit};
use avro_schema::schema::{
    Field as AvroField, Fixed, FixedLogical, IntLogical, LongLogical, Record, Schema as AvroSchema,
};
use error_stack::ResultExt;
use itertools::Itertools;

#[derive(derive_more::Display, Debug)]
#[display(fmt = "in field '{_0}'")]
struct FieldName(String);

/// Converts an arrow [`Schema`] to an Avro [`AvroSchema`].
pub fn to_avro_schema(schema: &Schema) -> error_stack::Result<AvroSchema, Error> {
    to_avro_record_schema(schema).map(AvroSchema::Record)
}

/// Converts an arrow [`Schema`] to an Avro [`Record`].
fn to_avro_record_schema(schema: &Schema) -> error_stack::Result<Record, Error> {
    let mut name_counter: i32 = 0;
    let fields: Vec<_> = schema
        .fields
        .iter()
        .map(|f| field_to_field(f, &mut name_counter))
        .try_collect()?;

    Ok(Record {
        name: "Record".to_string(),
        namespace: None,
        doc: None,
        aliases: vec![],
        fields,
    })
}

fn field_to_field(field: &Field, name_counter: &mut i32) -> error_stack::Result<AvroField, Error> {
    let schema = type_to_schema(field.data_type(), field.is_nullable(), name_counter)
        .attach_printable_lazy(|| FieldName(field.name().to_owned()))?;
    Ok(AvroField::new(field.name(), schema))
}

fn type_to_schema(
    data_type: &DataType,
    is_nullable: bool,
    name_counter: &mut i32,
) -> error_stack::Result<AvroSchema, Error> {
    // Nullable types are represented in Avro as the union of `null` and the type.
    Ok(if is_nullable {
        AvroSchema::Union(vec![
            AvroSchema::Null,
            type_to_schema_non_nullable(data_type, name_counter)?,
        ])
    } else {
        type_to_schema_non_nullable(data_type, name_counter)?
    })
}

fn record_name(name_counter: &mut i32) -> String {
    *name_counter += 1;
    format!("r{name_counter}")
}

fn type_to_schema_non_nullable(
    data_type: &DataType,
    name_counter: &mut i32,
) -> error_stack::Result<AvroSchema, Error> {
    Ok(match data_type {
        DataType::Null => AvroSchema::Null,
        DataType::Boolean => AvroSchema::Boolean,
        DataType::Int32 => AvroSchema::Int(None),
        DataType::Int64 => AvroSchema::Long(None),
        DataType::Float32 => AvroSchema::Float,
        DataType::Float64 => AvroSchema::Double,
        DataType::Binary => AvroSchema::Bytes(None),
        DataType::LargeBinary => AvroSchema::Bytes(None),
        DataType::Utf8 => AvroSchema::String(None),
        DataType::LargeUtf8 => AvroSchema::String(None),
        DataType::Struct(fields) => AvroSchema::Record(Record::new(
            record_name(name_counter),
            fields
                .iter()
                .map(|f| field_to_field(f, name_counter))
                .try_collect()?,
        )),
        DataType::Date32 => AvroSchema::Int(Some(IntLogical::Date)),
        DataType::Time32(TimeUnit::Millisecond) => AvroSchema::Int(Some(IntLogical::Time)),
        DataType::Time64(TimeUnit::Microsecond) => AvroSchema::Long(Some(LongLogical::Time)),
        DataType::Timestamp(TimeUnit::Millisecond, None) => {
            AvroSchema::Long(Some(LongLogical::LocalTimestampMillis))
        }
        DataType::Timestamp(TimeUnit::Microsecond, None) => {
            AvroSchema::Long(Some(LongLogical::LocalTimestampMicros))
        }
        DataType::Interval(IntervalUnit::MonthDayNano) => {
            let mut fixed = Fixed::new("", 12);
            fixed.logical = Some(FixedLogical::Duration);
            AvroSchema::Fixed(fixed)
        }
        DataType::FixedSizeBinary(size) => AvroSchema::Fixed(Fixed::new("", (*size) as usize)),
        other => error_stack::bail!(Error::Unimplemented(other.clone())),
    })
}

#[cfg(test)]
mod tests {
    use super::to_avro_schema;
    use arrow::datatypes::{DataType, Field, Schema};

    #[test]
    fn test_simple_struct_to_avro() {
        let arrow_schema = Schema::new(vec![
            Field::new("a", DataType::Int64, true),
            Field::new("b", DataType::Float64, false),
            Field::new("c", DataType::Utf8, true),
        ]);

        let avro_schema = to_avro_schema(&arrow_schema).unwrap();
        insta::assert_json_snapshot!(&avro_schema);
    }

    #[test]
    fn test_nested_struct_to_avro() {
        let arrow_schema = Schema::new(vec![
            Field::new("a", DataType::Int64, true),
            Field::new("b", DataType::Float64, false),
            Field::new("c", DataType::Utf8, true),
            Field::new(
                "d",
                DataType::Struct(vec![
                    Field::new("a", DataType::Int64, true),
                    Field::new("b", DataType::Float64, false),
                ]),
                false,
            ),
        ]);

        let avro_schema = to_avro_schema(&arrow_schema).unwrap();
        insta::assert_json_snapshot!(&avro_schema);
    }
}
