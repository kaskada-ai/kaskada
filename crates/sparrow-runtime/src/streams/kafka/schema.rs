use schema_registry_converter::{
    async_impl::schema_registry::{get_schema_by_subject, SrSettings},
    schema_registry_common::SubjectNameStrategy,
};

use error_stack::{IntoReport, Result, ResultExt};

/// TODO: Pending Implementation
#[allow(unused)]
#[derive(Debug, derive_more::Display)]
pub enum Error {
    #[allow(dead_code)]
    AvroNotEnabled,
    AvroSchemaConversion,
    SchemaRequest,
    UnsupportedSchema,
}
impl error_stack::Context for Error {}

#[allow(unused)]
/// TODO: Pending implementation
pub async fn get_kafka_schema() -> Result<String, Error> {
    let sr_settings = SrSettings::new(String::from("http://localhost:8085"));
    let subject_name_strategy =
        SubjectNameStrategy::TopicNameStrategy(String::from("my-topic"), false);
    let schema = get_schema_by_subject(&sr_settings, &subject_name_strategy)
        .await
        .into_report()
        .change_context(Error::SchemaRequest)?;
    Ok(schema.schema)
}
