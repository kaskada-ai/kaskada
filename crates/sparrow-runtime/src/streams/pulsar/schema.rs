use arrow::array::ArrowPrimitiveType;
use arrow::datatypes::{Schema, SchemaRef, TimestampMicrosecondType};

use error_stack::{IntoReport, Result, ResultExt};
use serde;
use serde_json::Value;
use std::sync::Arc;

#[derive(Debug, derive_more::Display)]
pub enum Error {
    #[allow(dead_code)]
    AvroNotEnabled,
    AvroSchemaConversion,
    MalformedSchema,
    RoundTripError,
    SchemaRequest,
    UnsupportedSchema,
}
impl error_stack::Context for Error {}

pub fn get_output_schema(schema: SchemaRef) -> Result<SchemaRef, Error> {
    let fields = schema.fields();

    // Avro does not support certain types that we use internally for the implicit columns.
    // For the `_time` (timestamp_ns) column, we cast to timestamp_us, sacrificing nano precision.
    // (Optionally, we could provide a separate column composed of the nanos as a separate i64).
    // The `_subsort` and `_key_hash` columns are dropped.
    let time_us = Arc::new(arrow::datatypes::Field::new(
        "_time",
        TimestampMicrosecondType::DATA_TYPE,
        false,
    ));

    let mut new_fields = vec![time_us];
    fields
        .iter()
        .skip(3) // Skip the `_time`, `_subsort`, and `_key_hash` fields
        .for_each(|f| new_fields.push(f.clone()));

    Ok(Arc::new(Schema::new(new_fields)))
}

#[cfg(not(feature = "avro"))]
fn format_schema(_schema: SchemaRef) -> Result<String, Error> {
    error_stack::bail!(Error::AvroNotEnabled)
}

#[cfg(feature = "avro")]
pub fn format_schema(schema: SchemaRef) -> Result<String, Error> {
    let avro_schema = sparrow_arrow::avro::to_avro_schema(&schema)
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| {
            format!("failed to convert arrow schema to avro schema {schema}")
        })?;
    serde_json::to_string(&avro_schema)
        .into_report()
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| {
            format!("failed to serialize avro schema to json string: {avro_schema:?}")
        })
}

fn schema_from_formatted(formatted_schema: &str) -> Result<Schema, Error> {
    fn avro_schema_from_json(formatted_schema: &str) -> Result<avro_schema::schema::Schema, Error> {
        serde_json::from_str(formatted_schema)
            .into_report()
            .change_context(Error::AvroSchemaConversion)
            .attach_printable_lazy(|| {
                format!("failed to deserialize avro schema from json string: {formatted_schema}")
            })
    }

    fn parse_json(formatted_schema: &str) -> Result<Value, Error> {
        serde_json::from_str(formatted_schema)
            .into_report()
            .change_context(Error::AvroSchemaConversion)
            .attach_printable_lazy(|| format!("invalid json string: {formatted_schema}"))
    }

    fn get_required_field<'a>(json: &'a Value, field_name: &'a str) -> Result<String, Error> {
        let v = json.get(field_name).ok_or_else(|| {
            error_stack::report!(Error::MalformedSchema)
                .attach_printable(format!("missing field {}", field_name))
        })?;
        serde_json::to_string(v).map_err(|e| {
            error_stack::report!(Error::RoundTripError)
                .attach_printable(format!("failed to serialize json value: {e}"))
        })
    }

    fn convert_to_arrow(avro_schema: &avro_schema::schema::Schema) -> Result<Schema, Error> {
        sparrow_arrow::avro::from_avro_schema(avro_schema)
            .change_context(Error::AvroSchemaConversion)
            .attach_printable_lazy(|| format!("from_avro_schema({:?}) failed", &avro_schema))
    }

    // The json string should have fields name, type, properties, and schema.
    // The type field tells us how to deserialize the schema.
    let json = parse_json(formatted_schema)?;
    let schema_type = get_required_field(&json, "type")?.to_lowercase();

    // we support avro and key/value schema types.  in the key/value case, both key
    // and value are json documents containing avro schemas.  We will flatten these
    // to a single schema.  This is guaranteed to be okay for CDC, since each field
    // corresponds to a unique Cassandra column.
    let avro_schema = match schema_type.as_str() {
        "avro" => avro_schema_from_json(formatted_schema)?,
        "key_value" => {
            let k = get_required_field(&json, "key")?;
            let v = get_required_field(&json, "value")?;
            let key_schema = avro_schema_from_json(k.as_str())?;
            let value_schema = avro_schema_from_json(v.as_str())?;
            combine_avro_schemas(key_schema, value_schema)?
        }
        _ => {
            let r = error_stack::report!(Error::UnsupportedSchema)
                .attach_printable(format!("unsupported schema type: {schema_type}"));
            return Err(r);
        }
    };

    Ok(convert_to_arrow(&avro_schema)?)
}

fn combine_avro_schemas(
    key_schema: avro_schema::schema::Schema,
    value_schema: avro_schema::schema::Schema,
) -> Result<avro_schema::schema::Schema, Error> {
    let fields1 = match key_schema {
        avro_schema::schema::Schema::Record(r) => r.fields,
        _ => {
            let r = error_stack::report!(Error::UnsupportedSchema).attach_printable(format!(
                "expected key schema to be a Record variant, found {:?}",
                key_schema
            ));
            return Err(r);
        }
    };
    let fields2 = match value_schema {
        avro_schema::schema::Schema::Record(r) => r.fields,
        _ => {
            let r = error_stack::report!(Error::UnsupportedSchema).attach_printable(format!(
                "expected value schema to be a Record variant, found {:?}",
                value_schema
            ));
            return Err(r);
        }
    };

    let mut fields = vec![];
    fields.extend(fields1);
    fields.extend(fields2);
    let record = avro_schema::schema::Record {
        name: "combined key/value schema".to_string(),
        namespace: None,
        doc: None,
        aliases: vec![],
        fields,
    };

    Ok(avro_schema::schema::Schema::Record(record))
}

#[derive(serde::Deserialize, Debug)]
struct SchemaResponse {
    #[allow(dead_code)]
    version: u32,
    #[serde(rename = "type")]
    schema_type: String,
    #[allow(dead_code)]
    timestamp: u64,
    data: String,
    #[allow(dead_code)]
    properties: serde_json::Value,
}

// auth_params looks like
// token:xxx
pub fn pulsar_auth_token(auth_params: &str) -> Result<&str, Error> {
    // split on the first colon
    let mut parts = auth_params.splitn(2, ':');
    // verify first part is "token"
    let auth_type = parts.next();
    let Some(auth_type) = auth_type else {
        return Err(Error::SchemaRequest.into()).attach_printable("missing auth_type");
    };
    if auth_type != "token" {
        return Err(Error::SchemaRequest.into())
            .attach_printable_lazy(|| format!("auth type {:?}", auth_type));
    }
    let auth_token = parts.next();
    let Some(auth_token) = auth_token else {
        return Err(Error::SchemaRequest.into()).attach_printable("missing auth token");
    };
    Ok(auth_token)
}

// retrieve the schema for the given topic via the admin api, with a REST call.
// we can't use the pulsar client because the schema is not exposed there in the Rust client.
pub async fn get_pulsar_schema(
    admin_service_url: &str,
    tenant: &str,
    namespace: &str,
    topic: &str,
    auth_params: &str,
) -> Result<Schema, Error> {
    let url = format!(
        "{}/admin/v2/schemas/{}/{}/{}/schema",
        admin_service_url, tenant, namespace, topic
    );
    tracing::debug!("requesting schema from {}", url);
    let client = reqwest::Client::new();
    let mut request_builder = client.get(&url);

    // Auth is generally recommended, but some local builds may be run
    // without auth for exploration/testing purposes.
    if !auth_params.is_empty() {
        request_builder = request_builder.header(
            reqwest::header::AUTHORIZATION,
            format!("Bearer {}", pulsar_auth_token(auth_params)?),
        )
    }

    let text = request_builder
        .send()
        .await
        .into_report()
        .change_context(Error::SchemaRequest)?
        .text()
        .await
        .into_report()
        .change_context(Error::SchemaRequest)?;

    let schema_response: SchemaResponse = serde_json::from_str(&text)
        .into_report()
        .change_context(Error::AvroSchemaConversion)
        .attach_printable_lazy(|| format!("from_str({:?}) failed", &text))?;
    if schema_response.schema_type != "AVRO" {
        return Err(Error::UnsupportedSchema.into())
            .attach_printable_lazy(|| format!("schema type {:?}", schema_response.schema_type));
    }
    let schema = schema_from_formatted(&schema_response.data)?;

    Ok(schema)
}
