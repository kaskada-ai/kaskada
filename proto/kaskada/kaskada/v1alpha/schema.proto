syntax = "proto3";

package kaskada.kaskada.v1alpha;

import "google/protobuf/empty.proto";

// Protobuf representation of the Fenl DataType.
//
// NOTE: Some of these types (especially related to time) are subject to
// change.
message DataType {
  oneof kind {
    // Primitive Fenl types.
    //
    // Each primitive enum corresponds to a specific Fenl type, as noted.
    PrimitiveType primitive = 1;

    // Struct types.
    //
    // The schema contains the field and types of the struct.
    //
    // String representation is `{ field: type, field: type, ... }`.
    Schema struct = 2;

    // Fenl type `window`.
    //
    // These cannot be output nor should they appear within a plan -- they are
    // generally compiled away as part of the implementation of aggrgeations.
    //
    // However, the `window` type may appear as the result of a view, allowing
    // the view to be used as part of aggregations.
    google.protobuf.Empty window = 3;

    // A list of a different type.
    DataType list = 4;

    // A map<k, v> type.
    Map map = 5;
  }

  message Map {
    string name = 1;
    bool ordered = 2;

    string key_name = 3;
    DataType key_type = 4;
    bool key_is_nullable = 5;

    string value_name = 6;
    DataType value_type = 7;
    bool value_is_nullable = 8;
  }

  enum PrimitiveType {
    PRIMITIVE_TYPE_UNSPECIFIED = 0;

    // Fenl type `null`.
    PRIMITIVE_TYPE_NULL = 1;
    // Fenl type `bool`.
    PRIMITIVE_TYPE_BOOL = 2;
    // Fenl type `i8`.
    PRIMITIVE_TYPE_I8 = 3;
    // Fenl type `i16`.
    PRIMITIVE_TYPE_I16 = 4;
    // Fenl type `i32`.
    PRIMITIVE_TYPE_I32 = 5;
    // Fenl type `i64`.
    PRIMITIVE_TYPE_I64 = 6;
    // Fenl type `u8`.
    PRIMITIVE_TYPE_U8 = 7;
    // Fenl type `u16`.
    PRIMITIVE_TYPE_U16 = 8;
    // Fenl type `u32`.
    PRIMITIVE_TYPE_U32 = 9;
    // Fenl type `u64`.
    PRIMITIVE_TYPE_U64 = 10;
    // Fenl type `f16`.
    PRIMITIVE_TYPE_F16 = 11;
    // Fenl type `f32`.
    PRIMITIVE_TYPE_F32 = 12;
    // Fenl type `f64`.
    PRIMITIVE_TYPE_F64 = 13;
    // Fenl type `string`.
    PRIMITIVE_TYPE_STRING = 14;
    // Fenl type `large_string`.
    PRIMITIVE_TYPE_LARGE_STRING = 27;

    // Fenl type `interval_days`.
    // A time interval represented as calendar days and milliseconds.
    PRIMITIVE_TYPE_INTERVAL_DAY_TIME = 15;
    // Fenl type `interval_months`.
    // A time interval represented as calendar months.
    PRIMITIVE_TYPE_INTERVAL_YEAR_MONTH = 16;

    // Fenl type `duration_s`.
    // Duration represented as a number of seconds.
    PRIMITIVE_TYPE_DURATION_SECOND = 17;
    // Fenl type `duration_ms`.
    // Duration represented as a number of milliseconds.
    PRIMITIVE_TYPE_DURATION_MILLISECOND = 18;
    // Fenl type `duration_us`.
    // Duration represented as a number of microseconds.
    PRIMITIVE_TYPE_DURATION_MICROSECOND = 19;
    // Fenl type `duration_ns`.
    // Duration represented as a number of nanoseconds.
    PRIMITIVE_TYPE_DURATION_NANOSECOND = 20;

    // Fenl type `timestamp_s`.
    // Timestamp represented as a number of seconds since epoch.
    PRIMITIVE_TYPE_TIMESTAMP_SECOND = 21;
    // Fenl type `timestamp_ms`.
    // Timestamp represented as a number of milliseconds since epoch.
    PRIMITIVE_TYPE_TIMESTAMP_MILLISECOND = 22;
    // Fenl type `timestamp_us`.
    // Timestamp represented as a number of microseconds since epoch.
    PRIMITIVE_TYPE_TIMESTAMP_MICROSECOND = 23;
    // Fenl type `timestamp_ns`.
    // Timestamp represented as a number of nanoseconds since epoch.
    PRIMITIVE_TYPE_TIMESTAMP_NANOSECOND = 24;

    // Fenl type `json`
    // Represents a json object in fenl
    PRIMITIVE_TYPE_JSON = 25;

    // Currently unsupported in fenl functions.
    PRIMITIVE_TYPE_DATE32 = 26;
  }
}

// A schema for an Arrow Struct column or an entire RecordBatch / table.
//
// When used, should generally be converted to a SchemaRef once and then
// additional references to the corresponding Arc used.
message Schema {
  repeated Field fields = 1;

  message Field {
    string name = 1;
    DataType data_type = 2;
    bool nullable = 3;
  }
}
