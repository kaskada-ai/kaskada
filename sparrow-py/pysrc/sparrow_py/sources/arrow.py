"""Provide sources based on PyArrow, including Pandas and CSV."""
from __future__ import annotations
from io import BytesIO
from typing import Optional
from typing import Union

import pandas as pd
import pyarrow.csv
import pyarrow.parquet
import pyarrow.json
import pyarrow as pa

from .source import Source

class Pandas(Source):
    """Source reading data from Pandas dataframe."""
    def __init__(
        self, dataframe: pd.DataFrame, *, schema: Optional[pa.Schema] = None, **kwargs
    ) -> None:
        """
        Create a source reading Pandas DataFrames.

        Parameters
        ----------
        dataframe : pd.DataFrame
            The DataFrame to start from.
        **kwargs : dict, optional
            Additional keyword arguments to pass to the super class.
            Should include the required column names.

        See Also
        --------
        Source.__init__ : For required keyword arguments.
        """
        if schema is None:
            schema = pa.Schema.from_pandas(dataframe)
        super().__init__(schema, **kwargs)
        self.add_data(dataframe)

    def add_data(self, data: pd.DataFrame) -> None:
        """Add data to the source."""
        table = pa.Table.from_pandas(data, self._schema, preserve_index=False)
        for batch in table.to_batches():
            self._ffi_table.add_pyarrow(data)

class PyList(Source):
    """Source reading data from lists of dicts."""
    def __init__(
        self, rows: dict | list[dict], *, schema: Optional[pa.Schema] = None, **kwargs
    ) -> None:
        """
        Create a source reading from rows represented as dicts.

        Parameters
        ----------
        rows : dict | list[dict]
            One or more represented as dicts.
        schema : pa.Schema, optional
            The schema to use.
            If not provided, the schema will be inferred from the `csv_string`.
        **kwargs : dict, optional
            Additional keyword arguments to pass to the super class.
            Should include the required column names.

        See Also
        --------
        Source.__init__ : For required keyword arguments.
        """
        if schema is None:
            schema = pa.Table.from_pylist(rows).schema
        super().__init__(schema, **kwargs)

        self._convert_options = pyarrow.csv.ConvertOptions(
            column_types=schema
        )
        self.add_rows(rows)

    def add_rows(self, rows: dict | list[dict]) -> None:
        """Add data to the source."""
        if isinstance(rows, dict):
            rows = [rows]
        table = pa.Table.from_pylist(rows, schema=self._schema)
        for batch in table.to_batches():
            self._ffi_table.add_pyarrow(batch)


class CsvString(Source):
    """Source reading data from CSV strings using PyArrow."""

    def __init__(
        self, csv_string: str, *, schema: Optional[pa.Schema] = None, **kwargs
    ) -> None:
        """
        Create a CSV String Source.

        Parameters
        ----------
        csv_string : str
            The CSV string to start from.
        schema : pa.Schema, optional
            The schema to use.
            If not provided, the schema will be inferred from the `csv_string`.
        **kwargs : dict, optional
            Additional keyword arguments to pass to the super class.
            Should include the required column names.

        See Also
        --------
        Source.__init__ : For required keyword arguments.
        """
        if isinstance(csv_string, str):
            csv_string = BytesIO(csv_string.encode("utf-8"))
        if schema is None:
            schema = pa.csv.read_csv(csv_string).schema
            csv_string.seek(0)
        super().__init__(schema, **kwargs)

        self._convert_options = pyarrow.csv.ConvertOptions(
            column_types=schema
        )
        self.add_string(csv_string)

    def add_string(self, csv_string: str | BytesIO) -> None:
        """Add data to the source."""
        if isinstance(csv_string, str):
            csv_string = BytesIO(csv_string.encode("utf-8"))
        content = pa.csv.read_csv(csv_string, convert_options = self._convert_options)
        for batch in content.to_batches():
            self._ffi_table.add_pyarrow(batch)

class JsonlString(Source):
    """Source reading data from line-delimited JSON strings using PyArrow."""

    def __init__(
        self, json_string: str, *, schema: Optional[pa.Schema] = None, **kwargs
    ) -> None:
        """
        Create a JSON String Source.

        Parameters
        ----------
        json_string : str
            The line-delimited JSON string to start from.
        schema : pa.Schema, optional
            The schema to use.
            If not provided, the schema will be inferred from the `csv_string`.
        **kwargs : dict, optional
            Additional keyword arguments to pass to the super class.
            Should include the required column names.

        See Also
        --------
        Source.__init__ : For required keyword arguments.
        """
        json_string = BytesIO(json_string.encode("utf-8"))
        if schema is None:
            schema = pa.json.read_json(json_string).schema
            json_string.seek(0)
        super().__init__(schema, **kwargs)

        self._parse_options = pyarrow.json.ParseOptions(
            explicit_schema=schema
        )
        self.add_string(json_string)

    def add_string(self, json_string: str | BytesIO) -> None:
        """Add data to the source."""
        if isinstance(json_string, str):
            json_string = BytesIO(json_string.encode("utf-8"))
        batches = pa.json.read_json(json_string, parse_options = self._parse_options)
        for batch in batches.to_batches():
            self._ffi_table.add_pyarrow(batch)

class Parquet(Source):
    """Source reading data from Parquet files."""
    def __init__(
        self, path: str, *, schema: Optional[pa.Schema] = None, **kwargs
    ) -> None:
        """
        Create a Parquet source.

        Parameters
        ----------
        path : str
            The path to the Parquet file to add.
        schema : pa.Schema, optional
            The schema to use.
            If not provided, the schema will be inferred from the `csv_string`.
        **kwargs : dict, optional
            Additional keyword arguments to pass to the super class.
            Should include the required column names.

        See Also
        --------
        Source.__init__ : For required keyword arguments.
        """
        if schema is None:
            schema = pa.parquet.read_schema(path)
        super().__init__(schema, **kwargs)

        self.add_file(path)

    def add_file(self, path: str) -> None:
        """Add data to the source."""
        table = pa.parquet.read_table(
            path,
            schema = self._schema,
        )
        for batch in table.to_batches():
            self._ffi_table.add_pyarrow(batch)