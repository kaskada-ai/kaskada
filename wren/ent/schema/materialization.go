package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/dialect/entsql"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
	"github.com/google/uuid"

	v1alpha "github.com/kaskada-ai/kaskada/gen/proto/go/kaskada/kaskada/v1alpha"
)

// Materialization holds the schema definition for the Materialization entity.
type Materialization struct {
	ent.Schema
}

// Fields of the Materialization.
func (Materialization) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).Default(uuid.New),
		field.Time("created_at").Default(microsecondNow).Immutable(),
		field.String("name").Immutable(),
		field.String("description").Optional(),
		field.String("expression").Immutable(),
		field.Bytes("with_views").GoType(&v1alpha.WithViews{}).Immutable(),
		field.Bytes("destination").GoType(&v1alpha.Materialization_Destination{}).Immutable(),
		field.Bytes("schema").GoType(&v1alpha.Schema{}).Immutable(),
		field.Bytes("slice_request").GoType(&v1alpha.SliceRequest{}).Immutable(),
		field.Bytes("analysis").GoType(&v1alpha.Analysis{}).Immutable(),
		field.Int64("data_version_id"),
	}
}

// Edges of the Materialization.
func (Materialization) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("owner", Owner.Type).Ref("materializations").Unique().Required(),
		edge.To("dependencies", MaterializationDependency.Type).Annotations(entsql.Annotation{OnDelete: entsql.Cascade}),
	}
}

// Indexes of the Materialization.
func (Materialization) Indexes() []ent.Index {
	return []ent.Index{
		// ensure materialization name is unique per owner
		index.Fields("name").Edges("owner").Unique(),
	}
}
