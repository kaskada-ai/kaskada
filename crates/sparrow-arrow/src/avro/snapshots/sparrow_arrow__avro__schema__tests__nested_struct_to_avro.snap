---
source: crates/sparrow-arrow/src/avro/schema.rs
expression: "&avro_schema"
---
{
  "type": "record",
  "name": "Record",
  "fields": [
    {
      "name": "a",
      "type": [
        "null",
        "long"
      ]
    },
    {
      "name": "b",
      "type": "double"
    },
    {
      "name": "c",
      "type": [
        "null",
        "string"
      ]
    },
    {
      "name": "d",
      "type": {
        "type": "record",
        "name": "r1",
        "fields": [
          {
            "name": "a",
            "type": [
              "null",
              "long"
            ]
          },
          {
            "name": "b",
            "type": "double"
          }
        ]
      }
    }
  ]
}
